numpy/lib/function_base.py

```diff
@@ -3612,6 +3666,78 @@ def quantile(a, q, axis=None, out=None,
     the median if ``q=0.5``, the same as the minimum if ``q=0.0`` and the
     same as the maximum if ``q=1.0``.
 
+    In the simplest case, where all weights are integers,
+    the `weights` argument can be seen as repeating the data in the sample.
+    For example, if a = [1,2,3] and weights = [1,2,3], this will be identical
+    to the case where there are no weights and a = [1,2,2,3,3,3], with the
+    quantiles interpolated in the cumulative probability space of the
+    expanded array.
+
+    The algorithm is illustrated here:
+
+    >>> a = np.array([3, 5, 4])  # sorted
+    >>> q = [0.25, 0.5, 0.75]
+    >>> axis = 0
+    >>> weights = [1, 3, 2]
+
+    Upon expansion, and converting q to quantile space ([0, 1])::
+
+        value       3   4   4   4   5   5
+        cum_prob    0  .2  .4  .6  .8   1
+        q=0.25            4
+        q=0.5                 4
+        q=0.75                    4.75
+
+    returns [4, 4, 4.75].  Note that even though the number 4 comprises 50%
+    of the entries, its quantile band only spans [0.2, 0.6].  By the same
+    token, the number 5 spans [0.8, 1].  In the "transition band" [0.6, 0.8],
+    one must interpolate between the values 4 and 5.  Similarly for [0, 0.2].
+
+    In this method, the computation of weighted quantile uses
+    normalized cumulative weights, with boundary conditions and renormalization
+    to maintain consistency with the case where all weights are integers.
+
+    The method works when all weights are >= 1.  If a weight is < 1,
```

Comment:
If this method doesn't work in all cases, then don't use it? Prefer the general rule, not the special case that breaks down

Reply:
I'd rather think this method, broadly speaking, is already a very general case.   The _less_ general case would be

either
1.) all weights are positive integers (frequency weights)
or
2.) all weights are [0, 1] and sum to 1 (probability weights).

Which is a lot more restraining.  The current method only asks for positive real numbers for weights, and works for either frequency weights, probability weights, and somewhere in between.

Reply:
When I talked to my two officemates (both PhD students who use R extensively), they both said there's no standard way to compute weighted quantile.  In fact, Wikipedia says the same thing:
https://en.wikipedia.org/wiki/Percentile#Weighted_percentile

If you look at a few more places, say
http://support.sas.com/documentation/cdl/en/procstat/68142/HTML/default/viewer.htm#procstat_univariate_details14.htm
or
https://stats.stackexchange.com/questions/13169/defining-quantiles-over-a-weighted-sample

you could see that they differ mostly in how they set the quantile band edges, and how they interpolate between those edges.

What I propose here is a method that

1.) can handle any arbitrary positive weights
2.) returns canonical answers in any "asymptotic limit"

By "asymptotic limit" I mean a case where all weights are positive integers, and the canonical answers are obtained when one frequency-expands-and-sorts the array and sets the quantiles to go from 0 in the left-most value to 1 in the right most value.  This is the most intuitive example of weighted quantile and can be easily illustrated graphically.

To further drive my point, we can also look at this example from SAS:
https://blogs.sas.com/content/iml/2016/08/29/weighted-percentiles.html
The results therein are nothing more than the proposed method with a `interpolation="nearest"` flag:
```
>>> np.quantile([1, 1.9, 2.2, 3, 3.7, 4.1, 5], q=[0.2, 0.4, 0.6, 0.8], interpolation='nearest')
array([1.9, 2.2, 3.7, 4.1])

>>> np.quantile([1, 1.9, 2.2, 3, 3.7, 4.1, 5], q=[0.2, 0.4, 0.6, 0.8], weights=[0.25, 0.05, 0.15, 0.25, 0.15, 0.10, 0.05], interpolation='nearest')
array([1. , 2.2, 3. , 3.7])
```

Another example here:
https://stackoverflow.com/questions/21844024/weighted-percentile-using-numpy/29677616#29677616
```
>>> np.quantile([1, 2, 9, 3.2, 4], q=[0, 0.5, 1], weights=[2, 1, 2, 4, 1])
array([1. , 3.2, 9. ])
>>> np.quantile([1, 2, 9, 3.2, 4], q=[0, 0.5, 1])
array([1. , 3.2, 9. ])
```

In conclusion, I don't think there is any convention we must follow.  The proposed method, as far as I know, is the best option, if we care about those "asymptotic limits".